File: frontend/src/features/gestion_interoperabilidad_frontend/utils/xmi/relations/GeneralizationMapper.ts
import type { Relation } from '@/features/gestion_modelado/shared/types/types';
import type { IRelationMapper } from '../core/interfaces';
import { uid } from '../core/utils';

export class GeneralizationMapper implements IRelationMapper {
    toXml(relation: Relation): string {
        return `<UML:Generalization xmi.id="EAID_${relation.id}" subtype="EAID_${relation.sourceId}" supertype="EAID_${relation.targetId}" />`;
    }

    fromXml(element: Element, idMap: Record<string, string>): Relation | null {
        const subtype = element.getAttribute("subtype") || "";
        const supertype = element.getAttribute("supertype") || "";
        
        const internalSrc = idMap[subtype];
        const internalTgt = idMap[supertype];
        
        if (internalSrc && internalTgt) {
            return {
                id: uid(),
                type: 'inheritance',
                sourceId: internalSrc,
                targetId: internalTgt,
                version: 0,
                hash: ''
            };
        }
        return null;
    }
}

File: backend/modules/gestion_interoperabilidad/shared/xmi/relations/generalization_mapper.py
from .relation_base import RelationBaseMapper

class GeneralizationMapper(RelationBaseMapper):
    def get_relation_type(self, elem):
        return "inheritance"

    def get_xml_tag(self, rel_type):
        return "UML:Generalization"

    def get_ea_type(self, rel_type):
        return "Generalization"
        
    def apply_xml_attribs(self, attribs, rel):
        attribs["subtype"] = f"EAID_{rel.get('sourceId', '')}"
        attribs["supertype"] = f"EAID_{rel.get('targetId', '')}"
